File: templates/python/agents/example_agent.py
"""
ExampleAgent — starter template.
Replace the capability implementations with your own logic.
"""

from interfaces import IAgent, AgentRequest, AgentResponse
from discovery.local_discovery import LocalDiscovery
from interfaces.iagent_discovery import DiscoveryEntry, NetworkInfo, HealthStatus
from datetime import datetime, timezone


class ExampleAgent(IAgent):
    # ── ERC-8004 Identity ─────────────────────────────────────────────────
    agent_id     = "sentrix://agent/example"
    owner        = "0xYourWalletAddress"
    metadata_uri = "ipfs://QmYourMetadataHashHere"
    metadata     = {
        "name":        "ExampleAgent",
        "version":     "0.1.0",
        "description": "A starter Sentrix agent",
        "tags":        ["example", "starter"],
    }

    # ── Capabilities ──────────────────────────────────────────────────────
    def get_capabilities(self):
        return ["echo", "ping"]

    # ── Request handling ──────────────────────────────────────────────────
    async def handle_request(self, req: AgentRequest) -> AgentResponse:
        if not await self.check_permission(req.from_id, req.capability):
            return AgentResponse.error(req.request_id, "Permission denied")

        match req.capability:
            case "echo":
                return AgentResponse.success(req.request_id, {"echo": req.payload})
            case "ping":
                return AgentResponse.success(req.request_id, {"pong": True, "agentId": self.agent_id})
            case _:
                return AgentResponse.error(req.request_id, f'Unknown capability: "{req.capability}"')

    # ── Discovery ─────────────────────────────────────────────────────────
    async def register_discovery(self) -> None:
        registry = LocalDiscovery.get_instance()
        await registry.register(DiscoveryEntry(
            agent_id=self.agent_id,
            name="ExampleAgent",
            owner=self.owner,
            capabilities=self.get_capabilities(),
            network=NetworkInfo(protocol="http", host="localhost", port=8080),
            health=HealthStatus(
                status="healthy",
                last_heartbeat=datetime.now(timezone.utc).isoformat()
            ),
            registered_at=datetime.now(timezone.utc).isoformat(),
        ))
        print("[ExampleAgent] registered with discovery layer")

    async def unregister_discovery(self) -> None:
        await LocalDiscovery.get_instance().unregister(self.agent_id)


# ── Dev runner ────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    import asyncio

    async def main():
        agent = ExampleAgent()
        await agent.register_discovery()

        # Smoke test
        req = AgentRequest(request_id="test-001", from_id="0xCaller", capability="ping", payload={})
        resp = await agent.handle_request(req)
        print("Response:", resp.to_dict())

    asyncio.run(main())

File: templates/typescript/agents/ExampleAgent.ts
import { IAgent, AgentMetadata } from '../interfaces/IAgent';
import { AgentRequest }           from '../interfaces/IAgentRequest';
import { AgentResponse }          from '../interfaces/IAgentResponse';
import { DiscoveryEntry }         from '../interfaces/IAgentDiscovery';

/**
 * ExampleAgent — starter template.
 * Replace the capability implementations with your own logic.
 */
export class ExampleAgent implements IAgent {
  // ─── ERC-8004 Identity ────────────────────────────────────────────────────
  readonly agentId     = 'sentrix://agent/example';
  readonly owner       = '0xYourWalletAddress';
  readonly metadataUri = 'ipfs://QmYourMetadataHashHere';
  readonly metadata: AgentMetadata = {
    name:        'ExampleAgent',
    version:     '0.1.0',
    description: 'A starter Sentrix agent',
    tags:        ['example', 'starter'],
  };

  // ─── Capabilities ─────────────────────────────────────────────────────────
  getCapabilities(): string[] {
    return ['echo', 'ping'];
  }

  // ─── Request handling ─────────────────────────────────────────────────────
  async handleRequest(req: AgentRequest): Promise<AgentResponse> {
    // Optional: enforce permissions
    if (this.checkPermission) {
      const allowed = await this.checkPermission(req.from, req.capability);
      if (!allowed) {
        return { requestId: req.requestId, status: 'error', errorMessage: 'Permission denied' };
      }
    }

    switch (req.capability) {
      case 'echo':
        return {
          requestId: req.requestId,
          status:    'success',
          result:    { echo: req.payload },
          timestamp: Date.now(),
        };

      case 'ping':
        return {
          requestId: req.requestId,
          status:    'success',
          result:    { pong: true, agentId: this.agentId },
          timestamp: Date.now(),
        };

      default:
        return {
          requestId:    req.requestId,
          status:       'error',
          errorMessage: `Unknown capability: "${req.capability}"`,
        };
    }
  }

  // ─── Discovery ────────────────────────────────────────────────────────────
  async registerDiscovery(): Promise<void> {
    // TODO: swap LocalDiscovery for GossipDiscovery or OnChainDiscovery
    const { LocalDiscovery } = await import('../discovery/LocalDiscovery');
    const registry = LocalDiscovery.getInstance();
    await registry.register({
      agentId:      this.agentId,
      name:         this.metadata.name,
      owner:        this.owner,
      capabilities: this.getCapabilities(),
      network:      { protocol: 'http', host: 'localhost', port: 8080, tls: false },
      health:       { status: 'healthy', lastHeartbeat: new Date().toISOString(), uptimeSeconds: 0 },
      registeredAt: new Date().toISOString(),
    });
    console.log(`[ExampleAgent] registered with discovery layer`);
  }

  async unregisterDiscovery(): Promise<void> {
    const { LocalDiscovery } = await import('../discovery/LocalDiscovery');
    await LocalDiscovery.getInstance().unregister(this.agentId);
  }

  // ─── ANR / Identity exposure ──────────────────────────────────────────────
  getAnr(): DiscoveryEntry {
    return {
      agentId:      this.agentId,
      name:         this.metadata.name,
      owner:        this.owner,
      capabilities: this.getCapabilities(),
      network:      { protocol: 'http', host: 'localhost', port: 8080, tls: false },
      health:       { status: 'healthy', lastHeartbeat: new Date().toISOString(), uptimeSeconds: 0 },
      registeredAt: new Date().toISOString(),
      metadataUri:  this.metadataUri,
    };
  }

  async getPeerId(): Promise<string | null> {
    // ExampleAgent has no signing key — swap for LocalKeystoreIdentity to get a real PeerId.
    return null;
  }

  // ─── Permissions ──────────────────────────────────────────────────────────
  async checkPermission(_caller: string, _capability: string): Promise<boolean> {
    // TODO: implement ERC-8004 delegation checks
    return true; // open by default for development
  }
}

// ── Entry point ───────────────────────────────────────────────────────────────
//
// Run directly:
//   npx ts-node agents/ExampleAgent.ts
//   SENTRIX_PORT=9090 npx ts-node agents/ExampleAgent.ts
//
// Or via sentrix-cli:
//   sentrix run ExampleAgent --port 8080
//
if (require.main === module) {
  (async () => {
    const { serve } = await import('../server');
    const port = parseInt(process.env.SENTRIX_PORT ?? '8080', 10);
    const agent = new ExampleAgent();
    await serve(agent, { port });
  })();
}
